pymc3/step_methods/slicer.py

```diff
@@ -44,33 +44,39 @@ def __init__(self, vars=None, w=1., tune=True, model=None, **kwargs):
         super(Slice, self).__init__(vars, [self.model.fastlogp], **kwargs)
 
     def astep(self, q0, logp):
-        self.w = np.resize(self.w, len(q0))
-        y = logp(q0) - nr.standard_exponential()
-
-        # Stepping out procedure
-        q_left = q0 - nr.uniform(0, self.w)
-        q_right = q_left + self.w
-
-        while (y < logp(q_left)).all():
-            q_left -= self.w
-
-        while (y < logp(q_right)).all():
-            q_right += self.w
-
-        q = nr.uniform(q_left, q_right, size=q_left.size)  # new variable to avoid copies
-        while logp(q) <= y:
-            # Sample uniformly from slice
-            if (q > q0).all():
-                q_right = q
-            elif (q < q0).all():
-                q_left = q
-            q = nr.uniform(q_left, q_right, size=q_left.size)
-
+        self.w = np.resize(self.w, len(q0))  # this is a repmat
+        q = np.copy(q0)  # TODO: find out if we need this
+        ql = np.copy(q0)  # l for left boundary
+        qr = np.copy(q0)  # r for right boudary
+        for i in range(len(q0)):
+            # uniformly sample from 0 to p(q), but in log space
+            y = logp(q) - nr.standard_exponential()
+            ql[i] = q[i] - nr.uniform(0, self.w[i])
+            qr[i] = q[i] + self.w[i]
+            # Stepping out procedure
+            while(y <= logp(ql)):  # changed lt to leq  for locally uniform posteriors
+                ql[i] -= self.w[i]
+            while(y <= logp(qr)):
+                qr[i] += self.w[i]
+
+            q[i] = nr.uniform(ql[i], qr[i])
+            while logp(q) < y:  # Changed leq to lt, to accomodate for locally flat posteriors
+                # Sample uniformly from slice
+                if q[i] > q0[i]:
+                    qr[i] = q[i]
+                elif q[i] < q0[i]:
+                    ql[i] = q[i]
+                q[i] = nr.uniform(ql[i], qr[i])
+
+            if self.tune:  # I was under impression from MacKays lectures that slice width can be tuned without
+                # breaking markovianness. Can we do it regardless of self.tune?(@madanh)
+                self.w[i] = self.w[i] * (self.n_tunes / (self.n_tunes + 1)) +\
+                    (qr[i] - ql[i]) / (self.n_tunes + 1)  # same as before
+            # unobvious and important: return qr and ql to the same point
+                qr[i] = q[i]
+                ql[i] = q[i]
         if self.tune:
```

Comment:
yes, but if it's not there then we will increment the tune counter even when not tuning, which is logically wrong and could detract somebody who's reading the code even more!